### PR TITLE
raftstore: Simplify raft local metrics

### PR DESCRIPTION
**Pull request by Connor1996**
Signed-off-by: Connor1996 <zbk602423539@gmail.com>

### What is changed and how it works?

Issue Number: Ref #12876
What's Changed:

```commit-message
Simplify raft local metrics by using local counter provided by rust-prometheus
```

### Check List 

Tests 

- No code

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by BusyJay** [comment]
What's the performance impact?

**Comment by Connor1996** [comment]
> What's the performance impact?

No difference, it doesn't involve the thread local storage. It's still the local metrics stored in `PollContext`, but we don't need to define the struct manually. It's already generated by `make_static_metircs!` with types of `LocalIntCounter`.

File: components/raftstore/src/store/local_metrics.rs [hunk 1]
```diff
@@ -123,317 +62,62 @@ impl RaftSendMessageMetrics {
...
+
+    pub leader_missing: Arc<Mutex<HashSet<u64>>>,
+
+    last_flush_time: Instant,
 }
 
 impl RaftMetrics {
     pub fn new(waterfall_metrics: bool) -> Self {
         Self {
+            ready: RaftReadyCounterVec::from(&STORE_RAFT_READY_COUNTER_VEC),
```
Comment (Connor1996) [inline components/raftstore/src/store/local_metrics.rs:105]: @BusyJay here is the key change. Before we define `RaftReadyMetrics` manually by repeating the fields in the static metrics definition. Now we use the local of the static metrics directly.

Comment (Connor1996) [reply components/raftstore/src/store/local_metrics.rs:105]: it's quite like calling `STORE_RAFT_READY_COUNTER_VEC.local()` but a static version

Comment (Connor1996) [reply components/raftstore/src/store/local_metrics.rs:105]: see https://github.com/tikv/rust-prometheus/blob/master/static-metric/examples/simple.rs#L24

Comment (BusyJay) [reply components/raftstore/src/store/local_metrics.rs:105]: Yes, my question is what's the difference between using provided APIs and defining it manually? Are they the exact same struct?

Comment (Connor1996) [reply components/raftstore/src/store/local_metrics.rs:105]: Yes, it's just a wrapper for `LocalIntCounter` inherently, see generated code https://github.com/tikv/rust-prometheus/blob/master/static-metric/examples/make_auto_flush_static_counter.rs#L270

File: components/raftstore/src/store/peer.rs [hunk 2]
```diff
@@ -1569,19 +1569,28 @@ where
         self.raft_group.snap()
     }
 
-    fn add_ready_metric(&self, ready: &Ready, metrics: &mut RaftReadyMetrics) {
-        metrics.message += ready.messages().len() as u64;
-        metrics.commit += ready.committed_entries().len() as u64;
-        metrics.append += ready.entries().len() as u64;
+    fn add_ready_metric(&self, ready: &Ready, metrics: &mut RaftMetrics) {
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:1572]: Why the change?

File: components/raftstore/src/store/worker/metrics.rs [hunk 3]
```diff
@@ -78,7 +78,7 @@ thread_local! {
     );
 }
 
-const METRICS_FLUSH_INTERVAL: u64 = 15_000; // 15s
+const METRICS_FLUSH_INTERVAL: u64 = 10_000; // 10s
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/metrics.rs:81]: Why?

Comment (Connor1996) [reply components/raftstore/src/store/worker/metrics.rs:81]: All other places use 10s

File: components/raftstore/src/store/peer.rs (hunk 2 above)
Comment (Connor1996) [reply components/raftstore/src/store/peer.rs:1572]: Don't want to introduce import from metrics as `RaftReadyMetrics` is deleted now.

**Review by tabokie — APPROVED** [review]
I hope we can run some microbenchmarks (on TiKV) for these changes. I see there're some in tests/benches, but there's no guide on which one we should run.

**Comment by Connor1996** [comment]
> I hope we can run some microbenchmarks (on TiKV) for these changes. I see there're some in tests/benches, but there's no guide on which one we should run.

No need cause there is no difference. It doesn't like other PRs which introduce additional overhead for collecting metrics. This PR just writes the code in a neater way.

**Comment by tabokie** [comment]
I'm not saying that targeting this specific PR. But as a general idea to avoid repeating the regression disaster in 6.2.

But for this PR, even if you are sure of it in theory, it is certainly better to have actual results as proof. Otherwise we wouldn't need CI because I'm sure we are all pretty confident of our own changes until proven wrong.

**Review by ethercflow — APPROVED** [review]
LGTM

**Comment by Connor1996** [comment]
> I'm not saying that targeting this specific PR. But as a general idea to avoid repeating the regression disaster in 6.2.
> 
> But for this PR, even if you are sure of it in theory, it is certainly better to have actual results as proof. Otherwise we wouldn't need CI because I'm sure we are all pretty confident of our own changes until proven wrong.

I agree it's better to take care of performance overhead to avoid repeating the regression. But not every PR worth to do the bench, it's a waste of the time(I cherish the memory of `/bench` bot).

**Comment by tabokie** [comment]
@Connor1996 That's why I'm talking about microbenchmarks. There're some many benches already but nobody seems to use them.

**Comment by Connor1996** [comment]
> @Connor1996 That's why I'm talking about microbenchmarks. There're some many benches already but nobody seems to use them.

I don't think the micro bench can find the previous regressions. It needs under heavy load. The daily bench with fine sensitivity would be a good practice.
